lib/foreman_tasks_core/runner/parent.rb

```diff
@@ -10,21 +10,23 @@ def initialize(targets = {}, suspended_action: nil)
       end
 
       def generate_updates
-        @outputs.reduce({}) do |acc, (key, value)|
+        base = {}
+        base[@suspended_action] = Runner::Update.new(ForemanTasksCore::ContinuousOutput.new, @exit_status) if @exit_status
```

Comment:
I'm sorry, but I can't find where the @suspended_action comes from for sure. Is it from dynflows `Action#suspended_action` ? isn't it an instance of `Action::Suspended`? In that case why are we using it as a key? :thinking:
What am I missing here?

Reply:
Each runner holds a reference to the suspended action which spawned the runner, that is `@suspended_action` in this context and it is indeed `Action::Suspended`. It gets passed to the runner on initialization and is stored in the top-level runner class here https://github.com/theforeman/foreman-tasks/blob/master/lib/foreman_tasks_core/runner/base.rb#L10 .

The return value from `generate_updates` is a hash, where keys are `Action::Suspended`-like objects, which serve as targets when sending events with updates to the per-host actions, and values are the updates themselves. Usually we want to send updates only to the per-host actions, but in case `@exit_status` is set, we need to send an update to the runner's action as well so it can finish.

Reply:
OH! now I see it, we are generating `Action::Suspended` from the host names :facepalm:
It makes me bit uneasy to have complicated objects as hash keys, I'd feel better if `Runner::Dispatcher` would know the `host -> execution_plan_id` mapping and we would just send hostnames and nil for global, but we can't always get what we wish for :) :shrug:

Your changes make it bit more readable as we can clearly see what are the keys supposed to look like at the beggining of the method so :+1: and thanks for the explanation